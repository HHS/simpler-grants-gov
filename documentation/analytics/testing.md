# Testing

## Running tests

All tests are automatically run as part of the [`ci-analytics.yml` GitHub Action](../../.github/workflows/ci-analytics.yml) that is triggered each time changes are pushed to a pull request which modifies the `analytics` package.

These CI checks calls `make test-audit` which runs the unit tests and validates that test coverage is above 80% for the `analytics` package. During development, you can also use the following commands to run all or part of the test suite.

### Running all tests with coverage

Run `make unit-test` from the root of the `analytics/` sub-directory to run all unit tests and print out a test coverage report.

### Running specific set of tests

To have greater control over which tests are run, you'll need to use `poetry run pytest` to access the pytest command line. The following commands show you how to invoke progressively smaller subsets of tests:

- `poetry run pytest` Runs all tests automatically collected in [test discovery](https://docs.pytest.org/en/stable/goodpractices.html#conventions-for-python-test-discovery)
- `poetry run pytest tests/datasets/` Runs all collected tests in the
  `analytics/tests/datasets/` sub-directory
- `poetry run pytest tests/dataset/test_base.py` Runs all collected the tests in
  the `analytics/tests/datasets/test_base.py` file
- `poetry run pytest tests/dataset/test_sprint_board.py::TestSprintBoard` Runs
  all tests in the `TestSprintBoard` class in `test_base.py`

Visit the pytest docs for additional information about [testing invocation and usage](https://docs.pytest.org/en/6.2.x/usage.html).

### Using pytest flags

Each of the test execution commands above can be modified or extended with a series of flags. Pytest documentation has [the full list of flags available](https://docs.pytest.org/en/stable/reference/reference.html#command-line-flags), but some commonly used flags include:

- `-x` Stands for "exit first" and exits test execution as soon as a test fails
- `--lf` Stands for "last fail" and runs only the tests that failed during the most recent test execution
- `-v` Stands for "verbose" and prints out the name of each test being executed and whether it passed or failed
- `-vv` Stands for "very verbose" and prints out detailed diffs between actual and expected outputs for failed tests

## Writing tests

[pytest](https://docs.pytest.org) is our test runner, which is simple but powerful. If you are new to pytest, reading up on how [fixtures work](https://docs.pytest.org/en/latest/explanation/fixtures.html) in particular might be helpful as it's one area that is a bit different than is common with other runners (and languages).

### Naming

pytest automatically discovers tests by [following a number of conventions](https://docs.pytest.org/en/stable/goodpractices.html#conventions-for-python-test-discovery)
(what it calls "collection").

For this project specifically:

- All tests live under `analytics/tests/`
- Under `tests/`, the organization mirrors the source code structure
  - The tests for [`analytics/src/analytics/datasets/`](../../analytics/src/analytics/datasets/)
    are found at [`analytics/tests/datasets/`](../../analytics/tests/datasets/)
- Integration tests have their own dedicated `tests/integrations/` testing sub-directory
- Create `__init__.py` files for each directory. This helps [avoid name conflicts
  when pytest is resolving tests](https://docs.pytest.org/en/stable/goodpractices.html#tests-outside-application-code).
- Test files should begin with the `test_` prefix, followed by the module the tests
  cover, for example, a file `foo.py` will have tests in a file `test_foo.py`.
- Test cases should begin with the `test_` prefix, followed by the function it's
  testing and some description of what about the function it is testing.
  - In `tests/datasets/test_base/test_base.py`, the `test_to_and_from_csv`
    function is a test (because it begins with `test_`).
  - Tests can be grouped in classes starting with `Test`, methods that start with
    `test_` will be picked up as tests, for example `TestFeature::test_scenario`.

There are occasions where tests may not line up exactly with a single source file, function, or otherwise may need to deviate from this exact structure, but this is the setup in general.

> [!TIP]
> Try to use test names that describe the expected behavior or the conditions you are testing. For example, `test_return_none_if_no_matching_sprint_exists()` is preferable to `test_get_sprint_name()`.
>
> If you can't easily describe the conditions or behavior you're testing, you may need to break your function up into multiple smaller tests. One common pattern is to create a test class (e.g. `TestGetSprintNameFromDate`) that represents the method or function you're testing and then create a test method for each test scenario (e.g. `test_return_previous_sprint_if_date_is_start_of_next_sprint()`)

### conftest files

`conftest.py` files are automatically loaded by pytest, making their contents available to tests without needing to be imported. They are an easy place to put shared test fixtures as well as define other pytest configuration (define hooks, load plugins, define new/override assert behavior, etc.).

They should never be imported directly.

The main `tests/conftest.py` holds widely useful fixtures included for all tests. Scoped `conftest.py` files can be created that apply only to the tests below them in the directory hierarchy, for example, the `tests/datasets/conftest.py` file would only be loaded for tests under `tests/datasets/`.

[More info about conftest files](https://docs.pytest.org/en/latest/how-to/fixtures.html?highlight=conftest#scope-sharing-fixtures-across-classes-modules-packages-or-session)


### Debugging tests

As the test suite begins to grow, running the full suite can take an increasingly long time to complete, which makes quickly iterating and debugging individual tests harder to do. In order to speed up iteration, a common workflow to follow when developing or testing a new feature includes:

1. Create a new test (or modify an existing test) add `assert 0` at the bottom of the test to force it to fail.
2. Run the test suite once with the optional `-x` flag to exit test execution when the new (or modified) test fails.
3. Iterate on the code and quickly test changes by using the `--lf` flag to rerun just the last failed tests.
4. Once the test checks for the behavior we expect *and* it runs successfully until `assert 0`, we know that we've implemented the new feature correctly.
5. Remove `assert 0` from the test then re-run the full test suite (e.g. `poetry run pytest`) to ensure that all tests pass and there are no regressions.
