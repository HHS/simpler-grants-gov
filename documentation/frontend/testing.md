# Testing

## E2E testing

### Spoofing logins

There are situations where we want to be able to test a "logged in" experience without having to script the test through the full login flow. In order to support this we have built a system to spoof the user login by placing a session cookie into the browser context.

This system is based on a `createSpoofedSessionCookie` which will create a client side cookie on the browser context that will effectively log in a fake user with the API.

Using this function, tests should work automatically in CI, but they will require a bit of manual setup to work locally.

#### Local setup

- run `make db-seed-local` in the /api directory. This will create the necessary DB records for the spoofed user and spit out an API auth token in a file at /api/e2e_token.tmp.
- copy the token variable declaration from the e2e_token.tmp file into your frontend .env.local file
- that's it! Running e2e tests using the functionality mentioned above should now work locally
