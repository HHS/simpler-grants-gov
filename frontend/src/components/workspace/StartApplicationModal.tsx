"use client";

import { useUser } from "src/services/auth/useUser";
import { startApplication } from "src/services/fetch/fetchers/clientApplicationFetcher";

import { useTranslations } from "next-intl";
import { useRouter } from "next/navigation";
import { useCallback, useRef, useState } from "react";
import { ModalRef, ModalToggleButton } from "@trussworks/react-uswds";

import { LoginModalBody } from "src/components/LoginModal";
import { SimplerModal } from "src/components/SimplerModal";
import { USWDSIcon } from "src/components/USWDSIcon";
import CompetitionStartFormIndividiual from "src/components/workspace/CompetitionStartFormIndividiual";

type StartApplicationModalProps = {
  competitionId: string;
  opportunityTitle: string;
};

const StartApplicationModal = ({
  competitionId,
  opportunityTitle,
}: StartApplicationModalProps) => {
  const modalRef = useRef<ModalRef>(null);
  const { user } = useUser();
  const router = useRouter();
  const t = useTranslations("OpportunityListing");
  const headerTranslation = useTranslations("HeaderLoginModal");
  const modalId = "start-application";
  const [validationError, setValidationError] = useState<string>();
  const [savedApplicationName, setSavedApplicationName] = useState<string>();
  const [error, setError] = useState<string>();
  const [loading, setLoading] = useState<boolean>();
  const token = user?.token || null;

  const handleSubmit = useCallback(() => {
    if (!token) {
      return;
    }
    if (validationError) {
      setValidationError(undefined);
    }
    if (!savedApplicationName) {
      setValidationError(t("startApplicationModal.validationError"));
      return;
    }
    setLoading(true);
    startApplication(savedApplicationName, competitionId, token)
      .then((data) => {
        const { applicationId } = data;
        router.push(`/workspace/applications/application/${applicationId}`);
      })
      .catch((error) => {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
        if (error.cause === "401") {
          setError(t("startApplicationModal.loggedOut"));
        } else {
          setError(t("startApplicationModal.error"));
        }
        console.error(error);
      });
  }, [competitionId, router, savedApplicationName, t, token, validationError]);

  const onClose = useCallback(() => {
    setError("");
    setLoading(false);
    setValidationError(undefined);
    setSavedApplicationName("");
  }, []);

  const onChange = useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
    setSavedApplicationName(e.target.value);
  }, []);

  return (
    <div className="display-flex flex-align-start">
      <ModalToggleButton
        modalRef={modalRef}
        data-testid={`open-start-application-modal-button`}
        opener
        className="usa-button"
      >
        <USWDSIcon name="add" />
        {t("startApplicationButtonText")}
      </ModalToggleButton>
      <SimplerModal
        modalRef={modalRef}
        className="text-wrap"
        modalId={modalId}
        titleText={
          token
            ? t("startApplicationModal.title")
            : t("startApplicationModal.login")
        }
        onKeyDown={(e) => {
          if (e.key === "Enter") handleSubmit();
        }}
        onClose={onClose}
      >
        {token ? (
          <CompetitionStartFormIndividiual
            opportunityTitle={opportunityTitle}
            loading={loading}
            error={error}
            onClose={onClose}
            onSubmit={handleSubmit}
            onChange={onChange}
            modalRef={modalRef}
            validationError={validationError}
          />
        ) : (
          <LoginModalBody
            helpText={headerTranslation("help")}
            buttonText={headerTranslation("button")}
            closeText={headerTranslation("close")}
            descriptionText={headerTranslation("description")}
            modalRef={modalRef}
          />
        )}
      </SimplerModal>
    </div>
  );
};

export default StartApplicationModal;
