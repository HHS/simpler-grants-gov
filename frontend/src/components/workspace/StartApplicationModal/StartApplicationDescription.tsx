import { ApplicantTypes } from "src/types/competitionsResponseTypes";
import { Organization } from "src/types/UserTypes";

import { useTranslations } from "next-intl";

export const StartApplicationDescription = ({
  organizations,
  applicantTypes,
}: {
  organizations: Organization[];
  applicantTypes: ApplicantTypes[];
}) => {
  const t = useTranslations(
    "OpportunityListing.startApplicationModal.description",
  );
  // individual
  if (!applicantTypes.includes("organization")) {
    return;
  }
  // ineligible
  if (!organizations.length && !applicantTypes.includes("individual")) {
    return (
      <div>
        <p>{t("organizationIntro")}</p>
        <ul>
          <li>{t("applyingForOrg")}</li>
          <li>{t("poc")}</li>
          <li>
            {t.rich("uei", {
              link: (chunk) => (
                <a
                  target="_blank"
                  rel="noopener noreferrer"
                  href="https://sam.gov"
                >
                  {chunk}
                </a>
              ),
            })}
          </li>
        </ul>
        <p>
          {t.rich("ineligibleGoToGrants", {
            link: (chunk) => (
              <a
                target="_blank"
                rel="noopener noreferrer"
                href="https://grants.gov"
              >
                {chunk}
              </a>
            ),
          })}
        </p>
      </div>
    );
  }
  // individual or organization
  if (applicantTypes.length === 2) {
    return (
      <div>
        <p>
          {t.rich("pilotGoToGrants", {
            link: (chunk) => (
              <a
                target="_blank"
                rel="noopener noreferrer"
                href="https://grants.gov"
              >
                {chunk}
              </a>
            ),
          })}
        </p>
        <p>{t("organizationIndividualIntro")}</p>
        <ul>
          <li>{t("poc")}</li>
          <li>
            {t.rich("uei", {
              link: (chunk) => (
                <a
                  target="_blank"
                  rel="noopener noreferrer"
                  href="https://sam.gov"
                >
                  {chunk}
                </a>
              ),
            })}
          </li>
        </ul>
      </div>
    );
  }
  // organization
  return (
    <div>
      <p>{t("organizationIntro")}</p>
      <ul>
        <li>{t("poc")}</li>
        <li>
          {t.rich("uei", {
            link: (chunk) => (
              <a
                target="_blank"
                rel="noopener noreferrer"
                href="https://sam.gov"
              >
                {chunk}
              </a>
            ),
          })}
        </li>
      </ul>
      <p>
        {t.rich("goToGrants", {
          link: (chunk) => (
            <a
              target="_blank"
              rel="noopener noreferrer"
              href="https://grants.gov"
            >
              {chunk}
            </a>
          ),
        })}
      </p>
    </div>
  );
};
