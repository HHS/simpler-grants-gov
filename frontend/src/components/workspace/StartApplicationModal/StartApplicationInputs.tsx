import { Organization } from "src/types/UserTypes";

import { useTranslations } from "next-intl";
import {
  ErrorMessage,
  Label,
  Select,
  TextInput,
} from "@trussworks/react-uswds";

export const StartApplicationOrganizationInput = ({
  onOrganizationChange,
  validationError = "",
  organizations,
  selectedOrganization,
}: {
  onOrganizationChange: (e: React.ChangeEvent<HTMLSelectElement>) => void;
  organizations: Organization[];
  selectedOrganization?: string; // organization_id
  validationError?: string;
}) => {
  const t = useTranslations(
    "OpportunityListing.startApplicationModal.fields.organizationSelect",
  );
  return (
    <>
      <Label
        id={"label-for-organization"}
        key={"label-for-organization"}
        htmlFor="application-organization"
        className="font-sans-2xs"
      >
        {t("label")}
      </Label>
      {validationError && <ErrorMessage>{validationError}</ErrorMessage>}

      <Select
        validationStatus={validationError ? "error" : undefined}
        id="create-application-organization-select"
        name="application-orgnization"
        onChange={onOrganizationChange}
        value={selectedOrganization || 0}
      >
        <option key={1} value={0} disabled>
          {t("default")}
        </option>
        {organizations.length &&
          organizations.map((organization) => (
            <option
              key={organization.organization_id}
              value={organization.organization_id}
            >
              {organization.sam_gov_entity.legal_business_name}
            </option>
          ))}
      </Select>
    </>
  );
};

export const StartApplicationNameInput = ({
  onNameChange,
  validationError = "",
}: {
  onNameChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  validationError?: string;
}) => {
  const t = useTranslations(
    "OpportunityListing.startApplicationModal.fields.name",
  );
  return (
    <>
      <Label
        id={"label-for-name"}
        key={"label-for-name"}
        htmlFor="application-name"
        className="font-sans-2xs"
      >
        {t("label")}
        <div>{t("description")}</div>
      </Label>
      {validationError && <ErrorMessage>{validationError}</ErrorMessage>}

      <TextInput
        validationStatus={validationError ? "error" : undefined}
        type="text"
        name="application-name"
        id="application-name"
        onChange={onNameChange}
      />
    </>
  );
};
