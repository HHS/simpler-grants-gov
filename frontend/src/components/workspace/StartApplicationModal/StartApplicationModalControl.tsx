"use client";

import { useClientFetch } from "src/hooks/useClientFetch";
import { useUser } from "src/services/auth/useUser";
import {
  ApplicantTypes,
  Competition,
} from "src/types/competitionsResponseTypes";
import { Organization } from "src/types/UserTypes";

import { useTranslations } from "next-intl";
import { useEffect, useRef, useState } from "react";
import { ModalRef, ModalToggleButton } from "@trussworks/react-uswds";

import { LoginModal } from "src/components/LoginModal";
import Spinner from "src/components/Spinner";
import { USWDSIcon } from "src/components/USWDSIcon";
import { StartApplicationModal } from "src/components/workspace/StartApplicationModal/StartApplicationModal";

type StartApplicationModalControlProps = {
  competitionId: string;
  opportunityTitle: string;
};

export const StartApplicationModalControl = ({
  competitionId,
  opportunityTitle,
}: StartApplicationModalControlProps) => {
  const modalRef = useRef<ModalRef>(null);
  const { user } = useUser();

  const { clientFetch: fetchUserOrganizations } = useClientFetch<
    Organization[]
  >("Error fetching user organizations");
  const { clientFetch: fetchCompetition } = useClientFetch<Competition>(
    "Error fetching competition",
  );
  const t = useTranslations("OpportunityListing.startApplicationModal");
  const headerTranslation = useTranslations("HeaderLoginModal");
  const [organizationsLoading, setOrganizationsLoading] = useState<boolean>();
  const [competitionsLoading, setCompetitionLoading] = useState<boolean>();
  const [competitionApplicantTypes, setCompetitionApplicantTypes] = useState<
    ApplicantTypes[]
  >([]);
  const [userOrganizations, setUserOrganizations] = useState<Organization[]>(
    [],
  );

  const token = user?.token || null;

  // TODO: these fetches should likely only happen if the user opens the modal, or if user is logged in so as not to overfetch
  // We could either move them around or just gate off of the presence of the token

  // see what the accepted applicant types are for this particular competition
  useEffect(() => {
    setCompetitionLoading(true);
    fetchCompetition(`/api/competitions/${competitionId}`)
      .then((competition) => {
        if (competition.open_to_applicants) {
          return setCompetitionApplicantTypes(competition.open_to_applicants);
        }
        console.error("Unable to find competition applicant designation");
      })
      .catch((e) => {
        console.error("Error fetching competition", e);
      })
      .finally(() => {
        setCompetitionLoading(false);
      });
  }, [competitionId, fetchCompetition]);

  // see what organizations the user belongs to, in order to handle organzation or
  // organization / individual competition types
  useEffect(() => {
    if (!token) {
      return;
    }
    setOrganizationsLoading(true);
    fetchUserOrganizations("/api/user/organizations", {
      cache: "no-store",
    })
      .then((organizations) => {
        return setUserOrganizations(organizations);
      })
      .catch((e) => {
        console.error("Error fetching user organizations", e);
      })
      .finally(() => {
        setOrganizationsLoading(false);
      });
  }, [fetchUserOrganizations, token]);

  return (
    <div className="display-flex flex-align-start">
      <ModalToggleButton
        modalRef={modalRef}
        data-testid="open-start-application-modal-button"
        opener
        disabled={organizationsLoading || competitionsLoading}
        className="usa-button"
      >
        {organizationsLoading || competitionsLoading ? (
          <>
            <Spinner className="height-105 width-105 button-icon-large" />{" "}
            {"Loading"}
          </>
        ) : (
          <>
            <USWDSIcon name="add" /> {t("startApplicationButtonText")}
          </>
        )}
      </ModalToggleButton>
      {token ? (
        <StartApplicationModal
          token={token}
          opportunityTitle={opportunityTitle}
          modalRef={modalRef}
          applicantTypes={competitionApplicantTypes}
          organizations={userOrganizations}
          loading={organizationsLoading || competitionsLoading || false}
          competitionId={competitionId}
        />
      ) : (
        <LoginModal
          helpText={headerTranslation("help")}
          buttonText={headerTranslation("button")}
          closeText={headerTranslation("close")}
          descriptionText={headerTranslation("description")}
          titleText={headerTranslation("title")}
          modalId="application-login-modal"
          modalRef={modalRef as React.RefObject<ModalRef>}
        />
      )}
    </div>
  );
};
