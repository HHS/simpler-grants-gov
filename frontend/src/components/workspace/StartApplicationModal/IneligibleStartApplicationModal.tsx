import { ApplicantTypes } from "src/types/competitionsResponseTypes";
import { Organization } from "src/types/UserTypes";

import { useTranslations } from "next-intl";
import { RefObject } from "react";
import {
  ModalFooter,
  ModalRef,
  ModalToggleButton,
} from "@trussworks/react-uswds";

import { SimplerModal } from "src/components/SimplerModal";
import { StartApplicationDescription } from "./StartApplicationDescription";

export const IneligibleApplicationStart = ({
  organizations,
  applicantTypes,
  modalRef,
  onClose,
  cancelText,
}: {
  organizations: Organization[];
  applicantTypes: ApplicantTypes[];
  modalRef: RefObject<ModalRef | null>;
  onClose: () => void;
  cancelText: string;
}) => {
  const t = useTranslations("OpportunityListing.startApplicationModal");
  return (
    <SimplerModal
      modalRef={modalRef}
      className="text-wrap maxw-tablet-lg font-sans-xs"
      modalId="start-application"
      titleText={t("ineligibleTitle")}
      onClose={onClose}
    >
      <StartApplicationDescription
        organizations={organizations}
        applicantTypes={applicantTypes}
      />
      <ModalFooter>
        <ModalToggleButton
          modalRef={modalRef}
          closer
          className="padding-105 text-center"
          onClick={onClose}
        >
          {cancelText}
        </ModalToggleButton>
      </ModalFooter>
    </SimplerModal>
  );
};
