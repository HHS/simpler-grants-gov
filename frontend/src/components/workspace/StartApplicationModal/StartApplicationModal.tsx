import { useClientFetch } from "src/hooks/useClientFetch";
import { ApplicantTypes } from "src/types/competitionsResponseTypes";
import { Organization } from "src/types/UserTypes";

import { useTranslations } from "next-intl";
import { useRouter } from "next/navigation";
import { RefObject, useCallback, useState } from "react";
import {
  Button,
  ErrorMessage,
  FormGroup,
  ModalFooter,
  ModalRef,
  ModalToggleButton,
} from "@trussworks/react-uswds";

import { SimplerModal } from "src/components/SimplerModal";
import { IneligibleApplicationStart } from "./IneligibleStartApplicationModal";
import { StartApplicationDescription } from "./StartApplicationDescription";
import {
  StartApplicationNameInput,
  StartApplicationOrganizationInput,
} from "./StartApplicationInputs";

export const StartApplicationModal = ({
  opportunityTitle,
  modalRef,
  applicantTypes,
  organizations,
  token,
  loading,
  competitionId,
}: {
  opportunityTitle: string;
  modalRef: RefObject<ModalRef | null>;
  applicantTypes: ApplicantTypes[];
  organizations: Organization[];
  token: string | null;
  loading: boolean;
  competitionId: string;
}) => {
  const t = useTranslations("OpportunityListing.startApplicationModal");
  const router = useRouter();
  const { clientFetch } = useClientFetch<{ applicationId: string }>(
    "Error starting application",
  );

  const [nameValidationError, setNameValidationError] = useState<string>();
  const [orgValidationError, setOrgValidationError] = useState<string>();
  const [savedApplicationName, setSavedApplicationName] = useState<string>();
  const [selectedOrganization, setSelectedOrganization] = useState<string>();
  const [error, setError] = useState<string>();
  const [updating, setUpdating] = useState<boolean>();

  const validateSubmission = useCallback((): boolean => {
    let valid = !!token;

    setOrgValidationError("");
    setNameValidationError("");

    if (!savedApplicationName) {
      setNameValidationError(t("fields.name.validationError"));
      valid = false;
    }
    if (!applicantTypes.includes("individual") && !selectedOrganization) {
      setOrgValidationError(t("fields.organizationSelect.validationError"));
      valid = false;
    }
    return valid;
  }, [token, savedApplicationName, applicantTypes, selectedOrganization, t]);

  const handleSubmit = useCallback(() => {
    const valid = validateSubmission();
    if (!valid) {
      return;
    }
    setUpdating(true);
    clientFetch("/api/applications/start", {
      method: "POST",
      body: JSON.stringify({
        applicationName: savedApplicationName,
        competitionId,
        organization: selectedOrganization,
      }),
    })
      .then((data) => {
        const { applicationId } = data;
        router.push(`/workspace/applications/application/${applicationId}`);
      })
      .catch((error) => {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
        if (error.cause === "401") {
          setError(t("loggedOut"));
        } else {
          setError(t("error"));
        }
        console.error(error);
      })
      .finally(() => {
        setUpdating(false);
      });
  }, [
    competitionId,
    router,
    savedApplicationName,
    t,
    selectedOrganization,
    validateSubmission,
    clientFetch,
  ]);

  const onClose = useCallback(() => {
    setError("");
    setUpdating(false);
    setNameValidationError("");
    setOrgValidationError("");
    setSavedApplicationName("");
  }, []);

  const onNameChange = useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
    setSavedApplicationName(e.target.value);
  }, []);

  const onOrganizationChange = useCallback(
    (e: React.ChangeEvent<HTMLSelectElement>) => {
      setSelectedOrganization(e.target.value);
    },
    [],
  );

  if (!organizations.length && !applicantTypes.includes("individual")) {
    return (
      <IneligibleApplicationStart
        modalRef={modalRef}
        cancelText={t("cancelButtonText")}
        onClose={onClose}
        organizations={organizations}
        applicantTypes={applicantTypes}
      />
    );
  }
  return (
    <SimplerModal
      modalRef={modalRef}
      className="text-wrap maxw-tablet-lg font-sans-xs"
      modalId={"start-application"}
      titleText={t("title")}
      onKeyDown={(e) => {
        if (e.key === "Enter") handleSubmit();
      }}
      onClose={onClose}
    >
      <StartApplicationDescription
        organizations={organizations}
        applicantTypes={applicantTypes}
      />
      <p className="font-sans-sm text-bold" data-testid="opportunity-title">
        {t("applyingFor")} {opportunityTitle}
      </p>
      <p className="font-sans-3xs">{t("requiredText")}</p>
      <FormGroup
        error={!!(nameValidationError || orgValidationError || error)}
        className="margin-top-1"
      >
        {applicantTypes.includes("organization") && (
          <StartApplicationOrganizationInput
            onOrganizationChange={onOrganizationChange}
            validationError={orgValidationError}
            organizations={organizations}
            selectedOrganization={selectedOrganization}
          />
        )}
        <StartApplicationNameInput
          validationError={nameValidationError}
          onNameChange={onNameChange}
        />
        {error && <ErrorMessage>{error}</ErrorMessage>}
      </FormGroup>
      <ModalFooter>
        <Button
          onClick={handleSubmit}
          type="button"
          data-testid="application-start-save"
          disabled={!!loading}
        >
          {loading || updating ? "Loading..." : t("saveButtonText")}
        </Button>
        <ModalToggleButton
          modalRef={modalRef}
          closer
          unstyled
          className="padding-105 text-center"
          onClick={onClose}
        >
          {t("cancelButtonText")}
        </ModalToggleButton>
      </ModalFooter>
    </SimplerModal>
  );
};
