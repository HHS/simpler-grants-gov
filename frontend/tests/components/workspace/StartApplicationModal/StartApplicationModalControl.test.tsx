import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { fakeCompetition, fakeOrganization } from "src/utils/testing/fixtures";
import { useTranslationsMock } from "src/utils/testing/intlMocks";

import { StartApplicationModalControl } from "src/components/workspace/StartApplicationModal/StartApplicationModalControl";

const mockUseUser = jest.fn(() => ({
  user: {
    token: "faketoken",
  },
}));

const mockRouterPush = jest.fn();
const mocks = {
  clientFetchMock: (url: string) => {
    if (url.match("competitions")) {
      return Promise.resolve(fakeCompetition);
    }
    return Promise.resolve([fakeOrganization]);
  },
};

jest.mock("next/navigation", () => ({
  useRouter: () => ({
    push: mockRouterPush,
  }),
}));

jest.mock("src/services/auth/useUser", () => ({
  useUser: () => mockUseUser(),
}));

jest.mock("next-intl", () => ({
  useTranslations: () => useTranslationsMock(),
}));

jest.mock("src/hooks/useClientFetch", () => ({
  useClientFetch: () => ({
    clientFetch: (url: string) => mocks.clientFetchMock(url),
  }),
}));

describe("StartApplicationModalControl", () => {
  beforeEach(() => {
    mockRouterPush.mockResolvedValue(true);
    mockUseUser.mockReturnValue({ user: { token: "a token" } });
  });
  afterEach(() => {
    jest.resetAllMocks();
  });

  it("matches snapshot", () => {
    const { container } = render(
      <StartApplicationModalControl
        competitionId="1"
        opportunityTitle="blessed opportunity"
      />,
    );
    expect(container).toMatchSnapshot();
  });
  it("modal can be opened and closed as expected", async () => {
    const { rerender } = render(
      <StartApplicationModalControl
        competitionId="1"
        opportunityTitle="blessed opportunity"
      />,
    );

    await waitFor(() => {
      expect(screen.queryByRole("dialog")).toHaveClass("is-hidden");
    });

    const toggle = await screen.findByTestId(
      "open-start-application-modal-button",
    );

    await userEvent.click(toggle);

    rerender(
      <StartApplicationModalControl
        competitionId="1"
        opportunityTitle="blessed opportunity"
      />,
    );

    await waitFor(() => {
      expect(screen.queryByRole("dialog")).not.toHaveClass("is-hidden");
    });

    const closeButton = await screen.findByText("cancelButtonText");

    await userEvent.click(closeButton);

    rerender(
      <StartApplicationModalControl
        competitionId="1"
        opportunityTitle="blessed opportunity"
      />,
    );
    await waitFor(() => {
      expect(screen.queryByRole("dialog")).toHaveClass("is-hidden");
    });
  });
  it("displays login modal on click if user is not logged in", async () => {
    mockUseUser.mockReturnValue({ user: { token: "" } });
    const { rerender } = render(
      <StartApplicationModalControl
        competitionId="1"
        opportunityTitle="blessed opportunity"
      />,
    );

    const toggle = await screen.findByTestId(
      "open-start-application-modal-button",
    );
    await userEvent.click(toggle);

    rerender(
      <StartApplicationModalControl
        competitionId="1"
        opportunityTitle="blessed opportunity"
      />,
    );

    await waitFor(() => {
      expect(screen.getByRole("dialog")).not.toHaveClass("is-hidden");
    });
    expect(screen.getByText("help")).toBeInTheDocument();
  });
  it("displays start application modal on click if user is logged in", async () => {
    const { rerender } = render(
      <StartApplicationModalControl
        competitionId="1"
        opportunityTitle="blessed opportunity"
      />,
    );

    const toggle = await screen.findByTestId(
      "open-start-application-modal-button",
    );

    await userEvent.click(toggle);

    rerender(
      <StartApplicationModalControl
        competitionId="1"
        opportunityTitle="blessed opportunity"
      />,
    );

    await waitFor(() => {
      expect(screen.getByText("requiredText")).toBeInTheDocument();
    });
    expect(screen.getByRole("dialog")).not.toHaveClass("is-hidden");
  });
  // as currently written this causes the useEffect to loops to run infinitely,
  // as the fetch function in the dependency gets constantly overwritten as part of the
  // implementation of the spy / mock. If we remove the fetch fn from the dependency
  // array, tests pass, but I'd rather have that in place and skip this test for now.
  // In the future we could make this more testable by encapsulating the onMount
  // fetch functionality into its own hook - DWS
  it.skip("calls fetch functions correctly", async () => {
    const spy = jest.spyOn(mocks, "clientFetchMock");

    render(
      <StartApplicationModalControl
        competitionId="1"
        opportunityTitle="blessed opportunity"
      />,
    );

    await waitFor(() => {
      expect(spy).toHaveBeenCalledTimes(2);
    });
    expect(spy).toHaveBeenCalledWith("/api/competitions/1");
    expect(spy).toHaveBeenCalledWith("/api/user/organizations", {
      cache: "no-store",
    });
  });
});
