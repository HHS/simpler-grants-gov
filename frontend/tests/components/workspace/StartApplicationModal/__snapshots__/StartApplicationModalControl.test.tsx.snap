// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StartApplicationModalControl matches snapshot 1`] = `
<div>
  <div
    class="display-flex flex-align-start"
  >
    <button
      aria-controls="start-application"
      class="usa-button usa-button"
      data-open-modal="true"
      data-testid="open-start-application-modal-button"
      disabled=""
      type="button"
    >
      <span
        aria-label="Loading!"
        class="grants-spinner height-105 width-105 button-icon-large"
        role="progressbar"
      />
       
      Loading
    </button>
  </div>
</div>
`;
